v3.2: Support all common XML node types (element, attribute, text, cdata) (#4592)
OAI/OpenAPI-Specification: src/oas.md

#### XML Object

A metadata object that allows for more fine-tuned XML model definitions.
When using a Schema Object with XML, if no XML Object is present, the behavior is determined by the XML Object's default field values.

##### Fixed Fields

| Field Name | Type | Description |
| ---- | :----: | ---- |

> | <a name="xml-node-type"></a>nodeType | `string` | One of `element`, `attribute`, `text`, `cdata`, or `none`, as explained under [XML Node Types](#xml-node-types).  The default value is `none` if `$ref`, `$dynamicRef`, or `type: "array"` is present in the [Schema Object](#schema-object) containing the XML Object, and `element` otherwise. |

Author: I'm not 100% the bits about `$ref` and `$dynamicRef` are backwards-compatible, because I'm not sure what, if anything, the behavior was in those cases before, as "property schema" was not well-defined.  If we mean "literal inline schema directly under a name under `properties` without references", then there was no defined behavior and we can do what we want.

Otherwise, it might not be, and we should let schemas containing `$ref` or `$dynamicRef` have a default `nodeType` of `element`, and folks can explicitly set them to `none`. I don't think that change would impact anything else in this PR, and might be the more conservative compatibility choice.

| <a name="xml-namespace"></a>namespace | `string` | The IRI ([[RFC3987]]) of the namespace definition. Value MUST be in the form of a non-relative IRI. |
| <a name="xml-prefix"></a>prefix | `string` | The prefix to be used for the [name](#xml-name). |
| <a name="xml-attribute"></a>attribute | `boolean` | Declares whether the property definition translates to an attribute instead of an element. Default value is `false`. If `nodeType` is present, this field MUST NOT be present.<br /><br />**Deprecated:** Use `nodeType: "attribute"` in place of `attribute: true` |
| <a name="xml-wrapped"></a>wrapped | `boolean` | MAY be used only for an array definition. Signifies whether the array is wrapped (for example, `<books><book/><book/></books>`) or unwrapped (`<book/><book/>`). Default value is `false`. The definition takes effect only when defined alongside `type` being `"array"` (outside the `items`). If `nodeType` is present, this field MUST NOT be present.<br /><br />**Deprecated:** Set `nodeType: "element"` explicitly in place of `wrapped: true` |

Note that when generating an XML document from object data, the order of the nodes is undefined.

> Use `prefixItems` to control node ordering.

Author: @ralfhandl there is an example... didn't you say you hadn't read the examples section yet? ;-)

In all seriousness, I'd love better examples once you get there, I struggle to come up with them.  But there is a `prefixItems` example.

> Use `prefixItems` to control node ordering.

Author: @ralfhandl I expanded those parts of the examples and added a link from this sentence so that it's less cryptic.

See [Appendix B](#appendix-b-data-type-conversion) for a discussion of converting values of various types to string representations.

This object MAY be extended with [Specification Extensions](#specification-extensions).

##### XML Node Types

> Each Schema Object describes a particular type of XML [node](https://dom.spec.whatwg.org/#interface-node) which is specified by the `nodeType` field, which has the following possible values.

Author: @ralfhandl I'd be happy to- what do I need to do for that?  If this is just standard respec I can read about in its docs, just lmk.

> Each Schema Object describes a particular type of XML [node](https://dom.spec.whatwg.org/#interface-node) which is specified by the `nodeType` field, which has the following possible values.

Author: @ralfhandl done- I made it normative so that the definitions (particularly the numbers which clarify which meaning of "text" is intended) are part of the spec.

Except for the special value `none`, these values have numeric equivalents in the DOM [specification](https://dom.spec.whatwg.org/#interface-node) which are given in parentheses after the name:

* `element` (1): The schema represents an element and describes its contents
* `attribute` (2): The schema represents an attribute and describes its value
* `text` (3): The schema represents a text node (parsed character data)
* `cdata` (4): The schema represents a CDATA section

> * `none`: The schema does not correspond to any node in the XML document, and its contents are included directly under the parent schema's node

Author: @ralfhandl Perhaps "the nodes corresponding to its subschema(s) are included directly under its parent schema's node"?

> * `none`: The schema does not correspond to any node in the XML document, and its contents are included directly under the parent schema's node

Author: I've added that wording, see if that works for you.

This also aligns with the inferred naming behavior defined under [XML Node Names](#xml-node-names).

To produce an element wrapping the list, set an explicit `nodeType: "element"` on the `type: "array"` schema.
When doing so, it is advisable to set an explicit name on either the wrapping element or the item elements to avoid them having the same inferred name.
See examples for expected behavior.

###### Implicit and Explicit `text` Nodes

If an `element` node has a primitive type, then the schema also produces an implicit `text` node described by the schema for the contents of the `element` node named by the property name (or `name` field).

Explicit `text` nodes are necessary if an element has both attributes and content.

Note that placing two `text` nodes adjacent to each other is ambiguous for parsing, and the resulting behavior is implementation-defined.

##### XML Node Names

The `element` and `attribute` node types require a name, which MUST be inferred from the schema as follows, unless overridden by the `name` field:

* For schemas directly under the [Components Object's](#components-object) `schemas` field, the component name is the inferred name.

> * For property schemas, and for array item schemas under a property schema, the property name is the inferred name

Reviewer: Do these two bullets need to be transposed to avoid an incompatible change? 

I assume that in the existing behavior the inferred node name for the `<Farm>.animals` property would be `animal`:

```yaml
components:
  schemas:
    Alien:
      type: string
    Farm:
      type: object
      properties:
        animal:
          $ref: "#/components/schemas/Alien"
```

###### XML Attribute, Prefix and Namespace

Note that the name of the root XML element comes from the component name.

```yaml
components:
  schemas:
    Person:
      type: object
      properties:
        id:
          type: integer
          format: int32
          xml:

>             attribute: true

Reviewer: @handrews just a reminder that you were thinking about changing this on the call.

>             attribute: true

Author: ```suggestion
            nodeType: attribute
```

</aliens>
```

###### Elements With Attributes And Text

```yaml
properties:
  animals:
    type: array
    xml:
      nodeType: element
      name: animals

>     items:
>       properties:
>         kind:
>           type: string
>           xml:
>             nodeType: attribute
>             name: animal

Reviewer 2: ```suggestion
    items:
      xml:
        name: animal
      properties:
        kind:
          type: string
          xml:
            nodeType: attribute
```
I think the items schema needs an xml field to set the name of the XML elements corresponding to the items.

</aliens>
```

###### Elements With Attributes And Text

```yaml
properties:
  animals:
    type: array
    xml:
      nodeType: element
      name: animals
    items:
      xml:
        name: animal
      properties:
        kind:
          type: string
          xml:
            nodeType: attribute
        content:
          type: string
          xml:
            nodeType: text
```

```xml
<animals>
  <animal kind="Cat">Fluffy</animals>
  <animal kind="Dog">Fido</animals>
<animals>
```

###### Referenced Element With CDATA

In this example, no element is created for the Schema Object that contains only  the `$ref`, as its `nodeType` defaults to `none`.
It is necessary to create a subschema for the CDATA section as otherwise the content would be treated as an implicit node of type `text`.

```yaml
paths:
  /docs:
    get:
      responses:
        "200":
          content:

>             application/xml:
>               $ref: "#/components/schemas/Documentation"

Author: ```suggestion
            application/xml:
              schema:
                $ref: "#/components/schemas/Documentation"
```

components:
  schemas:
    Documentation:
      type: object
      properties:
        content:
          type: string
          contentMediaType: text/html
          xml:
            nodeType: cdata
```

```xml
<Documentation>
  <![CDATA[<html><head><title>Awesome Docs</title></head><body></body><html>]]>
</Documentation>
```

Alternatively, the named root element could be set at the point of use and the root element disabled on the component:

```yaml
paths:
  /docs:
    get:
      responses:
        "200":
          content:
            application/xml:
              xml:
                nodeType: element
                name: StoredDocument
              $ref: "#/components/schemas/Documentation"
    put:
      requestBody:
        required: true
        content:

>           application/xml:
>             xml:
>               nodeType: element
>               name: UpdatedDocument
>             $ref: "#/components/schemas/Documentation"

Author: ```suggestion
          application/xml:
            schema:
              xml:
                nodeType: element
                name: UpdatedDocument
              $ref: "#/components/schemas/Documentation"
```

              nodeType: text
      - xml:
          name: Three
        type:
        - boolean
        - "null"
```

```xml
<OneTwoThree>
  <One>Some text</One>
  <Two unit="cubits">42</Two>
  <Three xsi:nil="true" />
</OneTwoThree>
```

The in-memory instance data that would produce the above XML snippet with the preceding schema would be:

```json
[
  "Some Text",
  {
    "unit": "cubits",
    "value": 42
  },
  null
]
```

In this next example, the `name` needs to be set for the element, while the `nodeType` needs to be set for the text nodes.

```yaml
components:
  schemas:
    Report:
      xml:
        nodeType: element
      type: array
      prefixItems:
      - xml:
          nodeType: text
        type: string
      - xml:
          name: data
        type: number
      - xml:
          nodeType: text
        type: string
```

```xml
<Report>Some preamble text.<data>42</data>Some postamble text.</Report>
```

The in-memory instance data structure for the above example would be:

```json
[
  "Some preamble text."
  42,
  "Some postamble text."
]
```

###### XML With `null` Values

Recall that the schema validates the in-memory data, not the XML document itself.

> The properties of the `"metadata"` element are omitted for brevity as it is here to show how the `null` value is represented.

Author: ```suggestion
The properties of the `"related"` element object are omitted for brevity as it is here to show how the `null` value is represented.
```
